src/upload/index.ts

```diff
@@ -0,0 +1,201 @@
+import {
+  HeadObjectCommand,
+  NotFound,
+  PutObjectCommand,
+  S3Client,
+} from "@aws-sdk/client-s3";
+import * as zlib from "zlib";
+import fs from "fs";
+
+import { LightBlockCache, lightBlockCache } from "@/cache";
+import { LightBlock } from "@/models/lightstreamer";
+import { logger } from "@/utils/logger";
+
+class UploadError extends Error {}
+
+export type BlockRange = {
+  start: number;
+  end: number;
+};
+
+export type BlockFile = {
+  file: string;
+  manifest: string;
+};
+
+export class LightBlockUpload {
+  private cache: LightBlockCache;
+  private s3Client: S3Client;
+  private chunkSizeMb: number;
+  private bucket: string;
+
+  constructor(cache: LightBlockCache) {
+    this.cache = cache;
+    if (!process.env["BUCKET_ENDPOINT"]) {
+      throw new UploadError("BUCKET_ENDPOINT not set");
+    }
+    if (!process.env["BUCKET_ACCESS_KEY_ID"]) {
+      throw new UploadError("BUCKET_ACCESS_KEY_ID not set");
+    }
+    if (!process.env["BUCKET_SECRET_ACCESS_KEY"]) {
+      throw new UploadError("BUCKET_SECRET_ACCESS_KEY not set");
+    }
+    if (!process.env["BUCKET_NAME"]) {
+      throw new UploadError("BUCKET_NAME not set");
+    }
+    if (!process.env["UPLOAD_CHUNK_SIZE_MB"]) {
+      throw new UploadError("UPLOAD_CHUNK_SIZE_MB not set");
+    }
+    this.chunkSizeMb = parseInt(
+      process.env["UPLOAD_CHUNK_SIZE_MB"] as string,
+      10,
+    );
+    this.bucket = process.env["BUCKET_NAME"];
+    this.s3Client = new S3Client({
+      region: "auto",
+      endpoint: process.env["BUCKET_ENDPOINT"],
+      credentials: {
+        accessKeyId: process.env["BUCKET_ACCESS_KEY_ID"],
+        secretAccessKey: process.env["BUCKET_SECRET_ACCESS_KEY"],
+      },
+    });
+  }
+
+  async upload(): Promise<void> {
+    // eslint-disable-next-line no-constant-condition
+    const fileName = "blocks";
+    const currentUploadSize = (await this.getFileSize(fileName)) || 0;
+    logger.info(
+      `Current file uploaded size: ${currentUploadSize}, creating new upload...`,
+    );
+    const files = await this.createBlockFiles(fileName, currentUploadSize);
+
+    logger.info(`Upload: begin...`);
+    const gzip = await this.gzipFile(files.file, `${files.file}.gz`);
```

Comment:
Have we considered other compression algorithms, zstd in particular?

Reply:
Speaking of compression: given that transactions contain mostly random data, I'd be curious to know if we actually gain anything substantial by compressing the blocks, and if it's worth offering it at all

Reply:
`gz` file is about half the size of non-gz.

Reply:
Relative to `gz` vs `zstd`, I just used `gz` due to its standard availability across platforms. I'll be honest I have never used zstd but it looks like a great compression algorithm when looking around.

Reply:
FWIW I had forgot to delete my local cache dir so I had some hex encoded string bytes that were bloating the gzip. When all the blocks are stored as binary these are the stats for compression:
binary: 212 MB
gz: 165 MB

So definitely not as good as 2x compression. Seems like it isn't worth it, I will remove it.